dcp: support DAOS

dsikich opened the pull request:
- Use mfu_file_t struct to wrap fd vs obj
    - create DAOS specific io functions
    - Create mfu_file_* functions that call DAOS backend
    - Add DAOS I/O functions to common library
    - call daos io backend in dcp
    - Add FindCART.cmake and FindDAOS.cmake
    - Add ifdef to DAOS specific code
    - create mfu_file in all tools for walk

Signed-off-by: Danielle Sikich <danielle.sikich@intel.com>

dsikich commented:
@adammoody So I fixed this one up, and I think it is much easier to see all of the changes from start to finish now.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -251,8 +506,33 @@ int mfu_hardlink(const char* oldpath, const char* newpath)
 /*****************************
- * Files
- ****************************/
+ *  * Files
```
There are a few spots like this where the editor is dropping in some extra * characters

dsikich replied:
@adammoody ok, will fix those.

adammoody commented:
Thanks, @dsikich .  Yes, this is much easier to read.  I'll build it and take a closer look at the POSIX portion over the next day or two.

adammoody commented on src/common/mfu_util.c:
```diff
@@ -36,6 +36,111 @@ mfu_loglevel mfu_debug_level = MFU_LOG_ERR;
+        free(global.iov_buf);
+}
+
+void daos_connect(int* rank, daos_handle_t* poh, daos_handle_t* coh,
```
There are some extra/missing indentation on various lines in this function.

dsikich replied:
@adammoody ok, fixed those.

adammoody commented on src/common/mfu_flist.h:
```diff
@@ -181,23 +181,26 @@ void mfu_flist_free(mfu_flist* flist);
     mfu_walk_opts_t* walk_opts, /* IN  - functions to perform during the walk */
-    mfu_flist flist             /* OUT - flist to insert walked items into */
+    mfu_flist flist,            /* OUT - flist to insert walked items into */
+    mfu_file_t* mfu_file        /* IN  - I/O filesystem functions to use during the walk */
```
Some of these functions take a list of source paths.  Will we need a corresponding mfu_file for each?  Or do we want to restrict all source paths to be of the same type?  Perhaps to start, we could restrict to one type, but in that case, we should probably run a check and kick out with an error if not.  I haven't looked at the whole PR yet, just thinking out loud...

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -257,7 +260,10 @@ static void walk_getdents_create(CIRCLE_handle* handle)
         /* stat top level item */
         struct stat st;
-        int status = mfu_lstat(path, &st);
+        int status;
+        mfu_file_t* mfu_file;
+        mfu_file = *pfile;
+        status = mfu_file_lstat(path, &st, mfu_file);
```
Let's put the declaration on the same line as initialization for these.
```suggestion
        mfu_file_t* mfu_file = *pfile;
        int status = mfu_file_lstat(path, &st, mfu_file);
```

adammoody replied:
Let's apply that change throughout.

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -506,10 +524,12 @@ static void walk_stat_process(CIRCLE_handle* handle)
     struct stat st;
-    int status = mfu_lstat(path, &st);
+    int status;
+    status = mfu_file_lstat(path, &st, mfu_file);
```
More thinking out loud.  Do we want to declare the mfu_file type here or deduce it from the path on the fly?

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -666,7 +691,7 @@ void mfu_flist_walk_paths(uint64_t num_paths, const char** paths,
             rate = ((double)all_count) / time_diff;
         }
-        MFU_LOG(MFU_LOG_INFO, "Walked %lu items in %f seconds (%f items/sec)",
+        MFU_LOG(MFU_LOG_INFO, "Walked %lu items in %f seconds (%f files/sec)",
```
Let's keep this as items, since some objects in the count might be directories or symlinks in addition to "files".
```suggestion
        MFU_LOG(MFU_LOG_INFO, "Walked %lu items in %f seconds (%f items/sec)",
```

adammoody commented on src/common/mfu_io.h:
```diff
@@ -33,6 +33,11 @@ extern "C" {
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <dirent.h>
+#ifdef DAOS_SUPPORT
```
```suggestion
#include <dirent.h>

#ifdef DAOS_SUPPORT
```

adammoody commented on src/common/mfu_io.h:
```diff
@@ -121,21 +165,14 @@ int mfu_fsync(const char* file, int fd);
 /* remove directory, retry a few times on EINTR or EIO */
 int mfu_rmdir(const char* dir);
 
-/* open directory, retry a few times on EINTR or EIO */
```
For organization, let's move the opendir, closedir calls back down to this "Directories" section.

adammoody commented on src/common/mfu_param_path.c:
```diff
@@ -418,6 +419,7 @@ void mfu_param_path_check_copy(uint64_t num, const mfu_param_path* paths,
     /* just have rank 0 check */
     if(rank == 0) {
         /* count number of readable source paths */
+/*
```
We'll want to find a DAOS equivalent for this I think, since it changes how the tools operate.

adammoody commented on src/common/mfu_param_path.c:
```diff
@@ -507,12 +510,12 @@ void mfu_param_path_check_copy(uint64_t num, const mfu_param_path* paths,
 
             /* check that dest is writable */
-            if(mfu_access(destpath->path, W_OK) < 0) {
+            /*if(mfu_access(destpath->path, W_OK) < 0) {
```
Same here

adammoody commented on src/common/mfu_param_path.c:
```diff
@@ -525,13 +528,13 @@ void mfu_param_path_check_copy(uint64_t num, const mfu_param_path* paths,
 
             /* check that parent is writable */
-            if(mfu_access(parent_str, W_OK) < 0) {
+            /*if(mfu_access(parent_str, W_OK) < 0) {
```
And here

adammoody commented on src/common/mfu_param_path.h:
```diff
@@ -114,7 +129,9 @@ char* mfu_param_path_copy_dest(
     const mfu_param_path* destpath, /* IN  - dest param path */
-    mfu_copy_opts_t* mfu_copy_opts  /* IN  - options to be used during copy */
+    mfu_copy_opts_t* mfu_copy_opts, /* IN  - options to be used during copy */
+    mfu_file_t* mfu_src_file,       /* IN  - I/O filesystem funcstions to use for copy of src */
+    mfu_file_t* mfu_dst_file        /* IN  - I/O filesystem funcstions to use for copy of dst */
```
```suggestion
    mfu_file_t* mfu_src_file,       /* IN  - I/O filesystem functions to use for copy of src */
    mfu_file_t* mfu_dst_file        /* IN  - I/O filesystem functions to use for copy of dst */
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -36,6 +36,111 @@ mfu_loglevel mfu_debug_level = MFU_LOG_ERR;
+
+        if (rank == 0) {
+                /* Get the global handle size. */
+                if (type == POOL_HANDLE)
+                        rc = daos_pool_local2global(*handle, &global);
+                else
+                        rc = daos_cont_local2global(*handle, &global);
```
```suggestion
                if (type == POOL_HANDLE) {
                        rc = daos_pool_local2global(*handle, &global);
                } else {
                        rc = daos_cont_local2global(*handle, &global);
                }
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -36,6 +36,111 @@ mfu_loglevel mfu_debug_level = MFU_LOG_ERR;
+
+        MPI_Bcast(&global.iov_buf_len, 1, MPI_UINT64_T, 0, MPI_COMM_WORLD);
+ 
+	global.iov_len = global.iov_buf_len;
+        global.iov_buf = malloc(global.iov_buf_len);
```
```suggestion
        global.iov_len = global.iov_buf_len;
        global.iov_buf = malloc(global.iov_buf_len);
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -36,6 +36,111 @@ mfu_loglevel mfu_debug_level = MFU_LOG_ERR;
+ 
+	global.iov_len = global.iov_buf_len;
+        global.iov_buf = malloc(global.iov_buf_len);
+        if (global.iov_buf == NULL)
+		MPI_Abort(MPI_COMM_WORLD, -1);
```
```suggestion
        if (global.iov_buf == NULL) {
                MPI_Abort(MPI_COMM_WORLD, -1);
        }
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -36,6 +36,111 @@ mfu_loglevel mfu_debug_level = MFU_LOG_ERR;
+		MPI_Abort(MPI_COMM_WORLD, -1);
+
+        if (rank == 0) {
+                if (type == POOL_HANDLE)
+                        rc = daos_pool_local2global(*handle, &global);
+                else
+                        rc = daos_cont_local2global(*handle, &global);
```
```suggestion
                if (type == POOL_HANDLE) {
                        rc = daos_pool_local2global(*handle, &global);
                } else {
                        rc = daos_cont_local2global(*handle, &global);
                }
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -36,6 +36,111 @@ mfu_loglevel mfu_debug_level = MFU_LOG_ERR;
+        MPI_Bcast(global.iov_buf, global.iov_buf_len, MPI_BYTE, 0, MPI_COMM_WORLD);
+
+        if (rank != 0) {
+                if (type == POOL_HANDLE)
+                        rc = daos_pool_global2local(global, handle);
+                else
+                        rc = daos_cont_global2local(*poh, global, handle);
```
```suggestion
                if (type == POOL_HANDLE) {
                        rc = daos_pool_global2local(global, handle);
                } else {
                        rc = daos_cont_global2local(*poh, global, handle);
                }
```

adammoody commented on src/common/mfu_util.c:
```diff
@@ -688,7 +793,7 @@ int mfu_compare_contents(
     mfu_progress* prg)             /* IN  - progress message structure */
 {
-    /* open source file */
+ /* open source file */
     int src_fd = mfu_open(src_name, O_RDONLY);
```
```suggestion
    /* open source file */
    int src_fd = mfu_open(src_name, O_RDONLY);
```

adammoody commented on src/common/mfu_util.h:
```diff
@@ -68,6 +70,14 @@ typedef enum {
 
+#ifdef DAOS_SUPPORT
+enum handleType {
+        POOL_HANDLE,
+        CONT_HANDLE,
+	ARRAY_HANDLE
```
```suggestion
        POOL_HANDLE,
        CONT_HANDLE,
        ARRAY_HANDLE
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -8,9 +8,91 @@
 #include "mpi.h"
 #include "libcircle.h"
-#include "mfu.h"
+#include "mfu.h" 
```
```suggestion
#include "mfu.h"
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -267,14 +415,99 @@ int main(int argc, char** argv)
             MFU_LOG(MFU_LOG_ERR, "Seconds in --progress must be non-negative: %d invalid", mfu_progress_timeout);
+            usage = 1;
         }
-        usage = 1;
```
Need to keep usage=1 here, or otherwise we'll deadlock.  All ranks need to know.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -8,9 +8,91 @@
+     * is not being copied from the root of the
+     * UNS path  */
+    int rc = 0;
+    char* src_path = argpaths[0];
+    char* dst_path = argpaths[1];
```
Can DAOS handle the user giving more than one source path, as in ```dcp src1 src2 dest/```?  If not, we should check for that and throw an error,

dsikich replied:
@adammoody No it will only support one source.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -267,14 +415,99 @@ int main(int argc, char** argv)
+                    if (rc != 0) {
+                        MFU_LOG(MFU_LOG_ERR, "Failed to create DFS2 container");
+                    }
+	            rc = daos_cont_open(src_poh, cuuid, DAOS_COO_RW, &dst_coh, &co_info, NULL);
```
```suggestion
                    rc = daos_cont_open(src_poh, cuuid, DAOS_COO_RW, &dst_coh, &co_info, NULL);
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -9,12 +9,187 @@
+#ifdef DAOS_SUPPORT
+#include <gurt/common.h>
+#include <gurt/hash.h>
+#endif
 #include "mfu.h"
```
```suggestion
#endif

#include "mfu.h"
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -57,7 +232,27 @@ int mfu_lchown(const char* path, uid_t owner, gid_t group)
 }
 
-/* calls chmod, and retries a few times if we get EIO or EINTR */
+int daos_chmod(const char *path, mode_t mode, mfu_file_t* mfu_file)
```
This is where the heavy lifting will be in this approach.  We either need to:
1) Create DAOS versions of the libc/posix functions so that they behave like posix functions with similar semantics, return code values, and setting of errno.  Then the code that makes calls to these will function assuming libc/posix.  For example, on chmod(2) it should return 0 on success, or -1 on failure and set errno appropriately.
2) Or we should define a new mfu_file_chmod replacement that by design does not follow libc/posix but instead has its own behavior that we define and control.  In that case, we'll need to review all code making this call and update it appropriately.

I don't have a good feel off hand for which way makes the most sense just yet,  I think option 1 might be a good start, but we can chat about it more.

dsikich replied:
@adammoody I'm thinking (1) is the better option, and we can discuss more of this tomorrow.

dsikich replied:
@adammoody The DFS API for DAOS does follow the POSIX return codes here: https://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c, so the wrappers would be returning the correct codes, matching the current mfu_ calls.

adammoody replied:
Excellent!  That will save a bunch of work then.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -437,7 +851,21 @@ int mfu_truncate(const char* file, off_t length)
+int daos_ftruncate(mfu_file_t* mfu_file, off_t length)
+{
+#ifdef DAOS_SUPPORT
+    int rc;
+    daos_off_t offset = (daos_off_t) length;
+    rc = dfs_punch(mfu_file->dfs, mfu_file->obj, offset, DFS_MAX_FSIZE);
```
```suggestion
    daos_off_t offset = (daos_off_t) length;
    int rc = dfs_punch(mfu_file->dfs, mfu_file->obj, offset, DFS_MAX_FSIZE);
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -558,6 +1028,37 @@ int mfu_rmdir(const char* dir)
+DIR* daos_opendir(const char* dir, mfu_file_t* mfu_file)
+{
+#ifdef DAOS_SUPPORT
+    struct dfs_mfu_t *dirp = NULL;
+    int rc;
+
+    dirp = calloc(1, sizeof(*dirp));
+    if (dirp == NULL)
+	    return NULL;
+
+    rc = dfs_lookup(mfu_file->dfs, dir, O_RDWR, &dirp->dir, NULL, NULL);
```
```suggestion
    struct dfs_mfu_t* dirp = calloc(1, sizeof(*dirp));
    if (dirp == NULL) {
	    return NULL;
    }

    int rc = dfs_lookup(mfu_file->dfs, dir, O_RDWR, &dirp->dir, NULL, NULL);
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -57,7 +252,27 @@ int mfu_lchown(const char* path, uid_t owner, gid_t group)
+    rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+    if (parent == NULL) {
+        fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+	return ENOENT;
```
It's a line like this one we want to look for.  We'd want to change this to something like:
```suggestion
	errno = ENOENT;
	return -1;
```

dsikich replied:
@adammoody I see, looks like the issue then is just with the wrappers around the DFS calls, so I can update these.

adammoody replied:
Maybe we don't need to set errno here, if we can rely on it being set from the dfs_lookup call.  Though we want to be sure to return -1 from ```mfu_file_chmod``` on any error, since code calling chmod might be doing something like:
```
int rc = mfu_file_chmod(...);
if (rc != -1)
  // success
else
  // error
```

dsikich replied:
@adammoody ok, I added checks in those mfu_file_* functions, if rc was not zero, I set the rc to return -1.

dsikich replied:
@adammoody I am going to update the changes we talked about yesterday in dcp: removing the daos short options, the check for the same pool, etc. Then I will add a check for the num of source paths.

adammoody replied:
> 
> 
> @adammoody ok, I added checks in those mfu_file_* functions, if rc was not zero, I set the rc to return -1.

I don't think all mfu_file_* functions return -1 on error.  We need to check the man page for each call to read what it should return on success and what it should return on failure.

dsikich replied:
@adammoody yeah, well all of the ones I updated say that in the man page (access, chmod, stat, Monod, close, truncate, and closer). The one that differs is lseek, which returns an offset on success, and -1 otherwise.

dsikich commented:
@adammoody I added a new commit to this PR. I thought it would be easier to see the changes for what we discussed in the call: Add checks for the number of source paths, verify return codes are correct, fix the same_pool check in dcp, and then adding the daos_ prefix to all of the DAOS command line inputs (like daos_src_pool, etc.).

dsikich commented:
@adammoody checking to see if you had a chance to look at this with the new changes?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -92,67 +100,78 @@ static mfu_copy_stats_t mfu_copy_stats;
             /* the file we're trying to open matches name and read/write mode,
              * so just return the cached descriptor */
-            return fd;
+            if (fd > 0) {
```
I think an fd==0 can be a valid open file descriptor, as well.
```suggestion
            if (fd >= 0) {
```
We should double check that it get initialized correctly.

dsikich replied:
@adammoody The fd value is initialized to -1 when the new mfu_file is created:

```C
/* return a newly allocated mfu_file structure, set default values on its fields */
mfu_file_t* mfu_file_new(void)
{
    mfu_file_t* mfile = (mfu_file_t*) MFU_MALLOC(sizeof(mfu_file_t));
    mfile->type       = POSIX;
    mfile->fd         = -1;
    mfile->only_daos  = false;
#ifdef DAOS_SUPPORT
    mfile->obj        = NULL;
    mfile->dfs        = NULL;
#endif
    return mfile;
}
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -92,67 +100,78 @@ static mfu_copy_stats_t mfu_copy_stats;
+#ifdef DAOS_SUPPORT
+        cache->obj  = mfu_file->obj;
+#endif
 #ifdef LUSTRE_SUPPORT
```
We should do some testing in an environment that has both Lustre and DAOS enabled.  For example, this code assumes the fd field is valid, but if it's dealing with a DAOS object, maybe that's not the case.  This particular code segment might be protected since the group lock is a lustre feature, but there might be other code like this out there.

Are you running on a system where you can have both enabled at once?

dsikich replied:
@adammoody No, I don't have an environment currently where I can run both unfortunately. The fd field should be initialized to -1 from what I saw, and as long as there is checking it shouldn't be an issue? Because if DAOS is enabled, the fd field will continue to stay -1?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -381,33 +399,27 @@ static int mfu_copy_acls(
           * We use the opaque method as we are not trying to alter the ACL contents.
           * Note that if the source is not a GPFS file-system, then the call will
           * fail with EINVAL and so we never try to apply this to the dest_path */
-     
```
Looks like about 5-6 empty lines got deleted.  Let's keep those in there since the spacing helps make it easier to read.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -416,21 +428,17 @@ static int mfu_copy_acls(
            bufsize = len + sizeof(struct gpfs_opaque_acl);
            MFU_LOG(MFU_LOG_DBG, "GPFS ACL buffer too small, needs to be %d",
                    (int) bufsize);
-     
```
Same down here, lost another 5-6 empty lines.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -755,9 +766,9 @@ static int mfu_copy_set_metadata_dirs(int levels, int minlevel, mfu_flist* lists
-                if (tmp_rc < 0) {
-                    rc = -1;
+                tmp_rc = mfu_copy_permissions(list, idx, dest, mfu_dst_file);
+                    if (tmp_rc < 0) {
```
some extra spaces find their way in here?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -855,7 +867,8 @@ static int mfu_create_directory(mfu_flist list, uint64_t idx,
     /* create the destination directory */
     MFU_LOG(MFU_LOG_DBG, "Creating directory `%s'", dest_path);
-    int mkdir_rc = mfu_mkdir(dest_path, DCOPY_DEF_PERMS_DIR);
+    int mkdir_rc;
+    mkdir_rc = mfu_file_mkdir(dest_path, DCOPY_DEF_PERMS_DIR, mfu_dst_file);
```
```suggestion
    int mkdir_rc = mfu_file_mkdir(dest_path, DCOPY_DEF_PERMS_DIR, mfu_dst_file);
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1303,6 +1320,7 @@ static int mfu_create_files(int levels, int minlevel, mfu_flist* lists,
 
             /* update number of files we have created for progress messages */
             mfu_progress_update(&total_count, create_prog);
+
```
Let's delete this line
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,46 +1574,80 @@ static int mfu_is_eof(const char* file, int fd)
-    void* buf = mfu_copy_opts->block_buf1;
+    /* declare daos vars */
+#ifdef DAOS_SUPPORT
+    daos_size_t		got_size;
+    d_sg_list_t		sgl;
+    d_iov_t		iov;
```
let's fix up tabs/spaces here

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,46 +1574,80 @@ static int mfu_is_eof(const char* file, int fd)
+    d_iov_t		iov;
+#endif
+
+    /* declare buffers */
+    size_t buf_size;
+    void* buf;
+    void* newbuf;
+    
+    /* set buffer and buffer size */
+    buf_size = mfu_copy_opts->block_size;
+    buf = mfu_copy_opts->block_buf1;
```
```suggestion
    /* set buffer and buffer size */
    size_t buf_size = mfu_copy_opts->block_size;;
    void* buf = mfu_copy_opts->block_buf1;
    void* newbuf;
```
Is newbuf used?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,46 +1574,80 @@ static int mfu_is_eof(const char* file, int fd)
+    buf = mfu_copy_opts->block_buf1;
+
+    /* set daos vars and buffers */
+#ifdef DAOS_SUPPORT
+    mfu_src_file->offset = (daos_off_t) offset;
+    mfu_dst_file->offset = (daos_off_t) offset;
+    sgl.sg_nr = 1;
+    D_ALLOC(buf, buf_size);
+    D_ASSERT(buf != NULL);
+    d_iov_set(&iov, buf, buf_size);
+    sgl.sg_iovs = &iov;
+    sgl.sg_nr_out = 1;
+    mfu_src_file->sgl = &sgl;
+    mfu_dst_file->sgl = &sgl;
+#endif
```
It looks like this code will execute regardless of whether the user is dealing with a DAOS object or not.  Let's only execute this if DAOS is being used.

Can you describe what the ```d_iov_set``` function does?

dsikich replied:
the iovec is actually defined in Cart/Gurt. It is setting the iov_buf and iov_len to the allocated buffer. 
```C
static inline void
d_iov_set(d_iov_t *iov, void *buf, size_t size)
{
        iov->iov_buf = buf;
        iov->iov_len = iov->iov_buf_len = size;
}

```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -2711,7 +2790,7 @@ static int mfu_fill_file(
     //mfu_copy_stats.total_size += (int64_t) total_bytes;
     //mfu_copy_stats.total_bytes_copied += (int64_t) total_bytes;
-
+    
```
```suggestion

```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -2900,10 +2979,10 @@ int mfu_flist_fill(mfu_flist list, mfu_copy_opts_t* mfu_copy_opts)
     return rc;
 }
-
+    
```
```suggestion

```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -3061,20 +3142,12 @@ int mfu_flist_file_sync_meta(mfu_flist src_list, uint64_t src_index, mfu_flist d
         }
     }
 
-    /* TODO: test ACLs and update if different */
```
It's ok to delete the commented code, but let's leave the TODO statement here.  That might still apply.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -3127,7 +3222,7 @@ mfu_copy_opts_t* mfu_copy_opts_new(void)
     opts->sparse        = false;
 
     /* Set default chunk size */
-    opts->chunk_size    = FD_CHUNK_SIZE;
+    opts->chunk_size    = 1*1024*1024;
```
Can we leave this as FD_CHUNK_SIZE, or was there a problem with that?

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -152,8 +153,10 @@ static void walk_getdents_process_dir(const char* dir, CIRCLE_handle* handle)
     /* TODO: may need to try these functions multiple times */
-    int fd = mfu_open(dir, O_RDONLY | O_DIRECTORY);
-    if (fd == -1) {
+    mfu_file_t* mfu_file;
+    mfu_file = *pfile;
```
```suggestion
    mfu_file_t* mfu_file = *pfile;
```

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -296,19 +300,23 @@ static void walk_getdents_process(CIRCLE_handle* handle)
 {
     /* TODO: may need to try these functions multiple times */
-    DIR* dirp = mfu_opendir(dir);
+    DIR* dirp;
+    mfu_file_t* mfu_file;
+    mfu_file = *pfile;
+    dirp = mfu_file_opendir(dir, mfu_file);
```
```suggestion
    mfu_file_t* mfu_file = *pfile;
    DIR* dirp = mfu_file_opendir(dir, mfu_file);
```

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -324,7 +332,8 @@ static void walk_readdir_process_dir(const char* dir, CIRCLE_handle* handle)
         while (1) {
             /* read next directory entry */
-            struct dirent* entry = mfu_readdir(dirp);
+            struct dirent* entry;
+            entry = mfu_file_readdir(dirp, mfu_file);
```
```suggestion
            struct dirent* entry = mfu_file_readdir(dirp, mfu_file);
```

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -409,7 +418,9 @@ static void walk_readdir_create(CIRCLE_handle* handle)
         /* stat top level item */
         struct stat st;
-        int status = mfu_lstat(path, &st);
+        mfu_file_t* mfu_file = *pfile;
+        int status = mfu_file_lstat(path, &st, mfu_file);
+
```
```suggestion
        mfu_file_t* mfu_file = *pfile;
        int status = mfu_file_lstat(path, &st, mfu_file);
```

adammoody replied:
Just deleting a line here to keep the code that checks a return code adjacent to where the return code value is created.

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -448,15 +459,19 @@ static void walk_readdir_process(CIRCLE_handle* handle)
 {
     /* TODO: may need to try these functions multiple times */
-    DIR* dirp = mfu_opendir(dir);
+    DIR* dirp;
+    mfu_file_t* mfu_file;
+    mfu_file = *pfile;
+    dirp = mfu_file_opendir(dir, mfu_file);
```
```suggestion
    mfu_file_t* mfu_file = *pfile;
    DIR* dirp = mfu_file_opendir(dir, mfu_file);
```

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -448,15 +459,19 @@ static void walk_readdir_process(CIRCLE_handle* handle)
         while (1) {
             /* read next directory entry */
-            struct dirent* entry = mfu_readdir(dirp);
+            struct dirent* entry;
+            entry = mfu_file_readdir(dirp, mfu_file);
```
```suggestion
            struct dirent* entry = mfu_file_readdir(dirp, mfu_file);
```

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -729,6 +756,8 @@ void mfu_flist_stat(
     /* step through each item in input list and stat it */
     uint64_t idx;
     uint64_t size = mfu_flist_size(input_flist);
+    mfu_file_t* mfu_file;
+    mfu_file = *pfile;
```
```suggestion
    mfu_file_t* mfu_file = *pfile;
```

adammoody commented on src/common/mfu_io.h:
```diff
@@ -34,6 +34,12 @@ extern "C" {
 #include <fcntl.h>
 #include <dirent.h>
 
+#ifdef DAOS_SUPPORT
+#include <daos.h>
+#include <daos_fs.h>
+#endif
+#include "mfu_param_path.h"
```
```suggestion
#ifdef DAOS_SUPPORT
#include <daos.h>
#include <daos_fs.h>
#endif

#include "mfu_param_path.h"
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -9,13 +9,203 @@
 #include <string.h>
 #include <errno.h>
 #include <stdarg.h>
+#include <assert.h>
+#include <libgen.h>
+#ifdef DAOS_SUPPORT
+#include <gurt/common.h>
+#include <gurt/hash.h>
+#endif
```
```suggestion
#include <assert.h>
#include <libgen.h>

#ifdef DAOS_SUPPORT
#include <gurt/common.h>
#include <gurt/hash.h>
#endif
```

dsikich commented on src/common/mfu_flist_copy.c:
```diff
@@ -92,67 +100,78 @@ static mfu_copy_stats_t mfu_copy_stats;
 
     /* cache the file descriptor */
-    if (newfd != -1) {
+    if (mfu_file->fd != -1) {
```
@adammoody no, it doesn't get set for a DAOS file. DAOS returns an open object handle when an object is opened of daos_handle_t type, so the fd field will stay -1 in that case.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -9,13 +9,203 @@
+}
+#endif
+
 /* calls access, and retries a few times if we get EIO or EINTR */
+int mfu_file_access(const char* path, int amode, mfu_file_t* mfu_file)
+{
+    int rc;
+    if (mfu_file->type == POSIX) {
+        rc = mfu_access(path, amode);
+    } else if (mfu_file->type == DAOS) {
+        rc = daos_access(path, amode);
+    }
+    if (rc != 0) {
+        rc = -1;
+    }
+    return rc;
+}
+
 int mfu_access(const char* path, int amode)
```
```suggestion
int mfu_file_access(const char* path, int amode, mfu_file_t* mfu_file)
{
    if (mfu_file->type == POSIX) {
        int rc = mfu_access(path, amode);
        return rc;
    } else if (mfu_file->type == DAOS) {
        int rc = daos_access(path, amode);
        if (rc != 0) {
            rc = -1;
        }
        return rc;
    } else {
        // let's catch this case and abort with an error
    }
}

/* calls access, and retries a few times if we get EIO or EINTR */
int mfu_access(const char* path, int amode)
```

adammoody replied:
Let's change the other mfu_file wrappers to look like this, where we add a catch all case and abort if they mfu_file->type is not known.  Also, let's just return the value from each function and then move the return code and errno to within the daos_access() code rather than straddle it here and there.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -57,7 +255,28 @@ int mfu_lchown(const char* path, uid_t owner, gid_t group)
+int daos_chmod(const char *path, mode_t mode, mfu_file_t* mfu_file)
+{
+#ifdef DAOS_SUPPORT
+    int rc;
+    dfs_obj_t *parent = NULL;
+    char *name = NULL, *dir_name = NULL;
+    parse_filename(path, &name, &dir_name);
+    assert(dir_name);
+    rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+    if (parent == NULL) {
+        fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+        errno = ENOENT;
+    }
+    rc = dfs_chmod(mfu_file->dfs, parent, name, mode);
+    if (rc) {
+        fprintf(stderr, "dfs_chmod %s failed (%d)\n", name, rc);
+        errno = rc;
+    }
+    return rc;
```
```suggestion
    char* name = NULL;
    char* dir_name = NULL;
    parse_filename(path, &name, &dir_name);
    assert(dir_name);

    dfs_obj_t* parent = NULL;
    int rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
    if (parent != NULL) {
        rc = dfs_chmod(mfu_file->dfs, parent, name, mode);
        if (rc) {
            MFU_LOG(MFU_LOG_ERR, "dfs_chmod %s failed (%d %s)", name, rc, strerror(rc));
            errno = rc;
            rc = -1;
        }
    } else {
        MFU_LOG(MFU_LOG_ERR, "dfs_lookup %s failed", dir_name);
        errno = ENOENT;
        rc = -1;
    }
    return rc;
```

adammoody replied:
What do you think of this?  Hopefully, I didn't mess something up there.  A few changes:

- avoid calling dfs_chmod if parent == NULL
- changed daos_chmod to return -1 on any error
- replaced fprintf with MFU_LOG
- a few style changes

adammoody commented on src/dcp/dcp.c:
```diff
@@ -202,6 +316,40 @@ int main(int argc, char** argv)
+#ifdef DAOS_SUPPORT
+            case 'x':
+	        rc = uuid_parse(optarg, src_pool_uuid);
+                if (rank == 0 && rc != 0) {
```
Some spacing/tab misalignment through here.

adammoody commented:
> 
> 
> @adammoody checking to see if you had a chance to look at this with the new changes?

Did a fresh sweep.  Mostly just polishing style at this point.  Let's update the mfu_file wrappers as I suggested in mfu_file_access and chmod to put all return code and errno handling within the daos_access/chmod wrappers, and ditto for the other functions.

dsikich commented:
> > @adammoody checking to see if you had a chance to look at this with the new changes?
> 
> Did a fresh sweep. Mostly just polishing style at this point. Let's update the mfu_file wrappers as I suggested in mfu_file_access and chmod to put all return code and errno handling within the daos_access/chmod wrappers, and ditto for the other functions.

@adammoody Awesome, thank you. I will ping you so you know when I have this updated. It should be sometime today or tomorrow.

dsikich commented:
@adammoody I updated the PR based on your review yesterday. I added them as three separate commits. One for mfu_flist_copy, one for mfu_flist_walk, and one for mfu_io. The last commit has all the changes to the mfu_file_* functions and the daos_ function wrappers. I can rebase them into one commit after you approve the PR. 

Do you think you will be able to look these over sometime this week?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -42,8 +42,13 @@
 #include <libgen.h> /* dirname */
+#include <stdbool.h>
 #include "libcircle.h"
 #include "dtcmp.h"
+#ifdef DAOS_SUPPORT
+#include <gurt/common.h>
+#include <gurt/hash.h>
+#endif
 #include "mfu.h"
```
```suggestion
#include "dtcmp.h"

#ifdef DAOS_SUPPORT
#include <gurt/common.h>
#include <gurt/hash.h>
#endif

#include "mfu.h"
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -92,67 +100,82 @@ static mfu_copy_stats_t mfu_copy_stats;
             /* the file we're trying to open matches name and read/write mode,
              * so just return the cached descriptor */
-            return fd;
+            if (fd != -1) {
+                return fd;
+            } 
+#ifdef DAOS_SUPPORT
+            else {
+                return obj;
+            }
+#endif
```
Since the return type has been changed from int to void, I think this probably just needs to ```return;```
```suggestion
            return;
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -497,23 +523,25 @@ static int mfu_copy_timestamps(
         /* if open for write, fsync */
         int read_flag = cache->read;
-        if (! read_flag) {
+        if (! read_flag && mfu_file->type == POSIX) {
```
Does DAOS have any kind of sync operation that would apply here?

dsikich replied:
@adammoody currently DAOS doesn't have a dfs_sync that is supported. It says it will be in the future though, but not sure when that will be.

dsikich replied:
@adammoody it is not yet supported as part of the DAOS filesystem API for POSIX yet at least.

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -296,19 +299,22 @@ static void walk_getdents_process(CIRCLE_handle* handle)
 {
     /* TODO: may need to try these functions multiple times */
-    DIR* dirp = mfu_opendir(dir);
+    DIR* dirp;
+    mfu_file_t* mfu_file = *pfile;
+    dirp = mfu_file_opendir(dir, mfu_file);
```
```suggestion
    mfu_file_t* mfu_file = *pfile;
    DIR* dirp = mfu_file_opendir(dir, mfu_file);
```

dsikich commented:
@adammoody I am just committing all of the style variable declarations you suggest as you go, I can rebase all the commits later. Since I know those won't change the functionality at all I am not going to run my tests against those.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -99,9 +339,37 @@ int mfu_utimensat(int dirfd, const char *pathname, const struct timespec times[2
+    } else {
+        parent = lookup_insert_dir(dir_name, mfu_file);
+    }
+    rc = dfs_stat(mfu_file->dfs, parent, name, buf);
+    if (rc) {
+        MFU_LOG(MFU_LOG_ERR, "dfs_stat %s failed (%d %s)",
+                name, rc, strerror(rc));
+        errno = rc;
+        rc = -1;
+    }
```
Should we avoid this call if the lookup for the parent fails?

dsikich replied:
@adammoody well you can call it in the case that the lookup is on the dfs root (ie "/"), which doesn't have a parent.

adammoody replied:
Let's add a comment about this, too, since it's another detail I'll probably forget :-)

adammoody commented on src/common/mfu_io.c:
```diff
@@ -297,13 +628,53 @@ int mfu_open(const char* file, int flags, ...)
-        /* if we still don't have a valid file, consider it an error */
-        if (fd < 0) {
-            /* we could abort, but probably don't want to here */
+       /* if we still don't have a valid file, consider it an error */
+       if (fd < 0) {
+           /* we could abort, but probably don't want to here */
+       }
```
```suggestion
         /* if we still don't have a valid file, consider it an error */
         if (fd < 0) {
             /* we could abort, but probably don't want to here */
         }
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -346,7 +737,52 @@ off_t mfu_lseek(const char* file, int fd, off_t pos, int whence)
+    daos_size_t got_size;
+    mfu_file->sgl->sg_iovs[0].iov_len = size;
+    int rc = dfs_read(mfu_file->dfs, mfu_file->obj, mfu_file->sgl, mfu_file->offset, &got_size, NULL); 
+    ssize_t size_read = (ssize_t)got_size;
```
This line needed?

dsikich replied:
@adammoody I suppose it can just return the got_size casted to the ssize_t type. I don't feel strongly about it either way.

adammoody replied:
It's fine to return this instead.  I just didn't see ```size_read``` being used anywhere, but maybe I missed that.  Let's drop it if it's not actually used.

dsikich replied:
@adammoody ok yeah I missed that. I removed that line.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -516,7 +1012,30 @@ void mfu_getcwd(char* buf, size_t size)
-/* create directory, retry a few times on EINTR or EIO */
+int daos_mkdir(const char* dir, mode_t mode, mfu_file_t* mfu_file) {
+#ifdef DAOS_SUPPORT
+    char* name     = NULL;
+    char* dir_name = NULL;
+
+    dfs_obj_t *parent = NULL;
+    parse_filename(dir, &name, &dir_name);
+    assert(dir_name);
+    int rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+
+    /* only call mkdir if directory doesn't already exist */
+    if (rc) {
```
```suggestion
    char* name     = NULL;
    char* dir_name = NULL;
    parse_filename(dir, &name, &dir_name);
    assert(dir_name);

    /* only call mkdir if directory doesn't already exist */
    dfs_obj_t *parent = NULL;
    int rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
    if (rc) {
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -558,6 +1092,35 @@ int mfu_rmdir(const char* dir)
+DIR* daos_opendir(const char* dir, mfu_file_t* mfu_file)
+{
+#ifdef DAOS_SUPPORT
+    struct dfs_mfu_t *dirp = NULL;
+    dirp = calloc(1, sizeof(*dirp));
```
```suggestion
    struct dfs_mfu_t *dirp = calloc(1, sizeof(*dirp));
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -558,6 +1092,35 @@ int mfu_rmdir(const char* dir)
+    if (rc) {
+        MFU_LOG(MFU_LOG_ERR, "dfs_lookup %s failed", dir);
+        errno = ENOENT;
+        return NULL;
```
I think we should also free the dirp that was calloc'd in this case

dsikich replied:
@adammoody it is freed in the daos_closedir function

adammoody replied:
Yeah, but in this error case, I don't think closedir would be called, since the opendir returned a failure.

dsikich replied:
@adammoody ah yeah ok I see what you mean

dsikich replied:
@adammoody added the free call

adammoody commented:
Thanks @dsikich .  The rework of the mfu_file wrappers looks good.

I did a quick scan of error checking from dfs_ routines, but I wasn't thorough with it.  It'd be good if you can double check those, especially the functions that have multiple dfs calls.

dsikich commented:
@adammoody Sure, do you mean the error codes that the dfs API returns? Or the error checking in the daos_ routines in mpifileutils? Most of the daos_ functions in mpifileutils call one or two dfs functions.

dsikich commented:
@adammoody should I go ahead and rebase this? Not sure if you want to review again after updates?

adammoody commented:
@dsikich , sure go ahead and rebase.  I can take a fresh look again after that.  Thanks for all of your work on this.

dsikich commented:
@adammoody Thank you.  I just rebased everything. Can you give it a fresh look?

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -50,6 +50,7 @@ static const char** CURRENT_DIRS;
 static flist_t* CURRENT_LIST;
 static int SET_DIR_PERMS;
 static int REMOVE_FILES;
+static mfu_file_t** pfile;
```
To make this more obvious that it's a global, let's rename ```pfile``` to something like ```CURRENT_PFILE```.

dsikich replied:
@adammoody ok done

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -539,29 +545,31 @@ static void walk_stat_process(CIRCLE_handle* handle)
+                          mfu_walk_opts_t* walk_opts, mfu_flist bflist,
+                          mfu_file_t* mfu_file)
 {
+
```
```suggestion
```

adammoody commented on src/common/mfu_flist_walk.c:
```diff
@@ -622,10 +630,13 @@ void mfu_flist_walk_paths(uint64_t num_paths, const char** paths,
         /* walk directories by calling stat on every item */
         CIRCLE_cb_create(&walk_stat_create);
         CIRCLE_cb_process(&walk_stat_process);
+        //        CIRCLE_cb_create(&walk_lustrestat_create);
+        //        CIRCLE_cb_process(&walk_lustrestat_process);
```
```suggestion
```

adammoody commented on src/common/mfu_param_path.c:
```diff
@@ -422,12 +431,12 @@ void mfu_param_path_check_copy(uint64_t num, const mfu_param_path* paths,
             else {
-                /* found a source path that we can't read, not fatal,
-                 * but print an error to notify user */
+                 /* found a source path that we can't read, not fatal,
+                  but print an error to notify user */
```
```suggestion
                /* found a source path that we can't read, not fatal,
                 * but print an error to notify user */
```

adammoody commented on src/common/mfu_param_path.h:
```diff
@@ -51,6 +52,20 @@ typedef struct mfu_param_path_t {
+    enum        {POSIX, DAOS} type;
+    int         fd;
+    bool        only_daos;
+    /* DAOS specific variables for I/O */
+#ifdef DAOS_SUPPORT
```
```suggestion
#ifdef DAOS_SUPPORT
    /* DAOS specific variables for I/O */
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -66,6 +148,12 @@ void print_usage(void)
     /* printf("  -g, --grouplock <id> - use Lustre grouplock when reading/writing file\n"); */
 #endif
     printf("  -b, --blocksize     - IO buffer size in bytes (default 1MB)\n");
+    printf("      --daos_src_pool      - DAOS source pool \n");
+    printf("      --daos_dst_pool      - DAOS destination pool \n");
+    printf("      --daos_src_cont      - DAOS source container \n");
+    printf("      --daos_dst_cont      - DAOS destination container \n");
+    printf("      --daos_svcl          - DAOS service level \n");
+    printf("      --daos_prefix        - DAOS prefix for unified namespace path \n");
```
```suggestion
    printf("      --daos-src-pool      - DAOS source pool \n");
    printf("      --daos-dst-pool      - DAOS destination pool \n");
    printf("      --daos-src-cont      - DAOS source container \n");
    printf("      --daos-dst-cont      - DAOS destination container \n");
    printf("      --daos-svcl          - DAOS service level \n");
    printf("      --daos-prefix        - DAOS prefix for unified namespace path \n");
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -114,6 +222,12 @@ int main(int argc, char** argv)
         {"blocksize"            , required_argument, 0, 'b'},
         {"debug"                , required_argument, 0, 'd'}, // undocumented
         {"grouplock"            , required_argument, 0, 'g'}, // untested
+        { "daos_src_pool"       , required_argument, 0, 'x'},
+        { "daos_dst_pool"       , required_argument, 0, 'D'},
+        { "daos_src_cont"       , required_argument, 0, 'y'},
+        { "daos_dst_cont"       , required_argument, 0, 'Y'},
+        { "daos_svcl"           , required_argument, 0, 'z'},
+        { "daos_prefix"         , required_argument, 0, 'X'},
```
```suggestion
        { "daos-src-pool"       , required_argument, 0, 'x'},
        { "daos-dst-pool"       , required_argument, 0, 'D'},
        { "daos-src-cont"       , required_argument, 0, 'y'},
        { "daos-dst-cont"       , required_argument, 0, 'Y'},
        { "daos-svcl"           , required_argument, 0, 'z'},
        { "daos-prefix"         , required_argument, 0, 'X'},
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -85,6 +173,22 @@ int main(int argc, char** argv)
     int rc = 0;
 
+    /* DAOS vars */ 
+#ifdef DAOS_SUPPORT
```
Since these are variables that will be set by the option processing, let's move those down to around to where the other variables like input are.

Also, let's initialize these so they are well defined with some value in the case the user does not set them through command line options.

adammoody replied:
I didn't check, but let's also be sure to order the options in the same way in the help message, their declaration in getopt, and the code that processes them.

dsikich replied:
@adammoody I verified they are in that order. It goes src pool, dest pool, src cont, dest cont, svcl, then the daos prefix.

adammoody commented:
Thanks, @dsikich , made another sweep.

Now that we're settling in on the options, would you also go ahead and update the doc/rst files for any commands where new command line options are added?

We've had cases where someone is dealing with two different file systems between source and destination or across multiple sources, including two different Lustre versions.  There are properties that we would be useful to capture for each those, and I'm thinking our new mfu_file_t object might be a good way to handle that.  We don't need to do anything in this PR, but I just wanted to give you a heads up on how this might change going forward.  Just thinking out loud.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -516,7 +1012,31 @@ void mfu_getcwd(char* buf, size_t size)
+
+    /* only call mkdir if the dir_name is not the root DFS directory */
+    if (strcmp(dir_name, "/") != 0) {
+        rc = dfs_mkdir(mfu_file->dfs, parent, name, S_IRWXU, 0);
```
Does this need a check to see if dfs_lookup succeeded?

dsikich replied:
@adammoody well it is ok if the lookup doesn't succeed, since not all directories in DFS have a parent. If there is an error then the check in the function will catch it.

adammoody replied:
Ok, let's add a comment about that, because I'll probably forget that detail later.

dsikich commented:
@adammoody Alright, I committed all the suggestions you made and updated based on your comments. I added the options to the dcp rst file. I have some more documentation that I wrote up for the DAOS portion of dcp, but I was hoping to put that in a separate PR? I did go ahead and add a short description of the options, but I also have a more comprehensive write-up I will add.

adammoody commented:
@dsikich , I merged a PR to fix a bug discovered when using O_DIRECT.

https://github.com/hpc/mpifileutils/pull/343/files

That likely produced the merge conflict on your PR.  Mind updating your PR to pick up those changes?

dsikich commented:
> @dsikich , I merged a PR to fix a bug discovered when using O_DIRECT.
> 
> https://github.com/hpc/mpifileutils/pull/343/files
> 
> That likely produced the merge conflict on your PR. Mind updating your PR to pick up those changes?

@dsikich yes, I just updated, thanks

dsikich commented:
@adammoody is there anything else I need to add to this PR?

adammoody commented:
@dsikich , I think it's probably good to go.  I'll take one more sweep tomorrow.

dsikich commented:
> @dsikich , I think it's probably good to go. I'll take one more sweep tomorrow.

@adammoody ok, thanks, sounds good

dsikich commented:
@adammoody did you get a chance to review yesterday?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1587,91 @@ static int mfu_is_eof(const char* file, int fd)
     size_t total_bytes = 0;
-    while(total_bytes < (size_t)length) {
+    char* buffer = mfu_copy_opts->block_buf1;
+    while(total_bytes <= (size_t)length) {
```
```suggestion
    while(total_bytes < (size_t)length) {
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1587,91 @@ static int mfu_is_eof(const char* file, int fd)
+            buffer = sgl.sg_iovs[0].iov_buf;
+        }
+#endif
+        num_of_bytes_read = mfu_file_read(src, buffer, left_to_read, mfu_src_file);
```
```suggestion
        ssize_t num_of_bytes_read = mfu_file_read(src, buffer, left_to_read, mfu_src_file);
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1587,91 @@ static int mfu_is_eof(const char* file, int fd)
             /* O_DIRECT requires particular read sizes */
             left_to_read = buf_size;
         }
+	ssize_t num_of_bytes_read;
```
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1587,91 @@ static int mfu_is_eof(const char* file, int fd)
             left_to_read = buf_size;
         }
+
         if(mfu_copy_opts->synchronous) {
             /* O_DIRECT requires particular read sizes */
             left_to_read = buf_size;
         }
```
I haven't thought about this change that I added with respect to what DAOS needed.  Is this new logic flow that computes the size of the read still valid for DAOS?

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -376,67 +393,71 @@ static int mfu_copy_acls(
 
     /* copy ACLs unless item is a link */
     if(type != MFU_TYPE_LINK) {
+
```
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -376,67 +393,71 @@ static int mfu_copy_acls(
-     
+
          /* check whether we read the ACL successfully */
+
```
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -376,67 +393,71 @@ static int mfu_copy_acls(
             * try to place it on dest_path */
            errno = 0;
            r = gpfs_putacl(dest_path, aclflags, aclbufmem);
+
```
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -639,12 +663,13 @@ static int mfu_copy_set_metadata(int levels, int minlevel, mfu_flist* lists,
                 }
             }
 
+
```
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1587,91 @@ static int mfu_is_eof(const char* file, int fd)
          * can read = max(buf size, remaining chunk) */
         size_t left_to_read = (size_t)length - total_bytes;
+        
         if(left_to_read > buf_size) {
+#ifdef DAOS_SUPPORT
+            if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
+                newbuf = mfu_copy_opts->block_buf1;
+  	        D_REALLOC(newbuf, buf, left_to_read);
+                d_iov_set(&iov, newbuf, left_to_read);
+            }
+#endif
             left_to_read = buf_size;
         }
```
Something here feels a bit off to me, but I might just have the wrong interpretation.  This if branch limits the buffer size to be no bigger than buf_size.  For example, coming in left_to_read might be 1GB but buf_size might only be 1MB.  This will reduce left_to_read down to 1MB, so that we only read from the file in chunks up to 1MB at a time.  However, the DAOS code is using the left_to_read value before it is reduced.  Should this code be moved closer to the read operation, after all of the manipulation to left_to_read has been completed?

dsikich replied:
@adammoody this should be resizing the DAOS buffer in the scatter gather list, which is used to tell DAOS where the data is in the DAOS read call below this one. I don't know what you mean about DAOS using the left_to_read before it is reduced?

dsikich replied:
@adammoody never mind, after giving this another look I see that resizing the DAOS buffer in the io vector is unnecessary. I am taking out that block.

adammoody replied:
The left_to_read can change values a few times before finally reaching the mfu_file_read the call.  So if buf_size=1MB, it might do something like:
```
// depending on the file size, this might be something really big like left_to_read=100GB
// or it might be something small like left_to_read=10 bytes
left_to_read = length - total_bytes;

// in case left_to_read is really big, this reduces it to be at most buf_size
if (left_to_read > buf_size)
  left_to_read = buf_size; // reduce left_to_read to 1MB

// this check is for cases where left_to_read < buf_size,
// because O_DIRECT requires particular values for the read count
if (synchronous)
  left_to_read = buf_size; // force left_to_read to be exactly 1MB
```

dsikich replied:
@adammoody yeah, thanks I appreciate it, I don't actually need to resize the io vector buffer there.

adammoody replied:
Does it work if you change this:
```
        if(left_to_read > buf_size) {
#ifdef DAOS_SUPPORT
            if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
                newbuf = mfu_copy_opts->block_buf1;
  	        D_REALLOC(newbuf, buf, left_to_read);
                d_iov_set(&iov, newbuf, left_to_read);
            }
#endif
            left_to_read = buf_size;
        }
```
to just:
```
        if(left_to_read > buf_size) {
            left_to_read = buf_size;
        }
```
and then add a d_iov_set line here:
```
#ifdef DAOS_SUPPORT
        if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
            d_iov_set(&iov, buf, left_to_read);
            buffer = sgl.sg_iovs[0].iov_buf;
        }
#endif
        num_of_bytes_read = mfu_file_read(src, buffer, left_to_read, mfu_src_file);
```

adammoody replied:
Note that I'm also dropping the D_REALLOC, which I'm guessing might not be necessary.  Above the while loop, buf is allocated to be buf_size bytes, and we never read more than that amount at a time.  We might read less, but I think it'd be fine to use a buffer that is bigger than is needed, so we don't have to allocate a new buffer.  We just need to set the size correctly on the scatter/gather list element.

dsikich replied:
@adammoody well I don't actually need to resize the buffer to left_to_read at all, since it can just read buf_size each time, which is fine. That was just a mistake on my part I think.

dsikich commented:
@adammoody thanks for your review! I took out the code block that was resizing the buffer, since it is fine for it to read buf_size each time. And I committed and then rebased all of the suggested commits you made regarding style/removing lines. I updated and tested again after removing that block of code as well.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1583,82 @@ static int mfu_is_eof(const char* file, int fd)
 
     /* write data */
     size_t total_bytes = 0;
+    char* buffer = mfu_copy_opts->block_buf1;
```
```suggestion
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1583,82 @@ static int mfu_is_eof(const char* file, int fd)
 
         /* read data from source file */
-        ssize_t num_of_bytes_read = mfu_read(src, in_fd, buf, left_to_read);
+#ifdef DAOS_SUPPORT
+        if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
+            buffer = sgl.sg_iovs[0].iov_buf;
+        }
+#endif
+        ssize_t num_of_bytes_read = mfu_file_read(src, buffer, left_to_read, mfu_src_file);
```
```suggestion
        ssize_t num_of_bytes_read = mfu_file_read(src, buf, left_to_read, mfu_src_file);
```

adammoody commented:
Thanks @dsikich , after that, I think we can simplify further by dropping the ```buffer``` variable, since there is only ```buf``` now.  I'll try to build and test this branch under a debugger next week for the POSIX part as a final pass.

dsikich commented:
> Thanks @dsikich , after that, I think we can simplify further by dropping the `buffer` variable, since there is only `buf` now. I'll try to build and test this branch under a debugger next week for the POSIX part as a final pass.

@adammoody yes good call that buffer variable is not needed either, I took it out.

dsikich commented:
@adammoody do you know if your next review will be later in the week? Just want to make sure I keep an eye on it around when you plan to review so that if there is anything I need to update I can do it fairly quickly.

adammoody commented:
@dsikich , thanks for staying responsive on this.  I plan to start on this today.  Depending on how it goes, I should be able to post an update later today or tomorrow.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -109,11 +201,39 @@ int main(int argc, char** argv)
         {"blocksize"            , required_argument, 0, 'b'},
         {"debug"                , required_argument, 0, 'd'}, // undocumented
         {"grouplock"            , required_argument, 0, 'g'}, // untested
+        { "daos-src-pool"       , required_argument, 0, 'x'},
+        { "daos-dst-pool"       , required_argument, 0, 'D'},
+        { "daos-src-cont"       , required_argument, 0, 'y'},
+        { "daos-dst-cont"       , required_argument, 0, 'Y'},
+        { "daos-svcl"           , required_argument, 0, 'z'},
+        { "daos-prefix"         , required_argument, 0, 'X'},
```
```suggestion
        {"daos-src-pool"       , required_argument, 0, 'x'},
        {"daos-dst-pool"       , required_argument, 0, 'D'},
        {"daos-src-cont"       , required_argument, 0, 'y'},
        {"daos-dst-cont"       , required_argument, 0, 'Y'},
        {"daos-svcl"           , required_argument, 0, 'z'},
        {"daos-prefix"         , required_argument, 0, 'X'},
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -202,6 +322,40 @@ int main(int argc, char** argv)
+            case 'x':
+	        rc = uuid_parse(optarg, src_pool_uuid);
+                if (rank == 0 && rc != 0) {
+                    MFU_LOG(MFU_LOG_ERR, "Failed to parse source pool uuid: '%s'", optarg);
```
I think we should be setting usage=1 on errors when processing the options.  Otherwise, I think it will try to execute anyway.  Be sure to set usage=1 on all ranks, though, while only rank 0 prints the error.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -202,6 +322,40 @@ int main(int argc, char** argv)
+    	        break;
+            case 'D':
+	        rc = uuid_parse(optarg, dst_pool_uuid);
+                if (rank == 0 && rc != 0) {
+                    MFU_LOG(MFU_LOG_ERR, "Failed to parse dst pool uuid: '%s'", optarg);
+                }
```
```suggestion
                if (rc != 0) {
                    if (rank == 0) {
                        MFU_LOG(MFU_LOG_ERR, "Failed to parse dst pool uuid: '%s'", optarg);
                    }
                    usage = 1;
                }
```

adammoody replied:
Also, does it make sense to have this option also imply ```mfu_src_file->type = DAOS;```, or not?

adammoody commented on src/dcp/dcp.c:
```diff
@@ -271,10 +425,97 @@ int main(int argc, char** argv)
+    char** argpaths = (&argv[optind]);
+
+#ifdef DAOS_SUPPORT
+    rc = daos_init();
```
To be extra safe, let's add an ```if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS)``` check around the code in this section so that it's not executed unless someone specified DAOS.

dsikich replied:
@adammoody well that will mess up the case where the unified namespaces are used, because I cannot fill in the types until after daos is initialized and then running some daos code on it to check if it is a "daos path"

dsikich replied:
@adammoody that would be ok if the src and dest pools/cont uuids were always passed in by cmd line, but that isn't always the case. You can also create POSIX DAOS containers that are linked to regular POSIX paths, and in that case you have to run DAOS code to lookup the uuids to check if it is actually within DAOS.

adammoody replied:
Ah, I see.  That makes sense.

adammoody replied:
We might be able to use something like stat or statfs in the future to help.  As part of the v0.11 release, I plan to look at a way to identify the underlying file system for each path.  This check will likely go in the mfu_param_path_set_all function.  We need that for things like finding the Lustre version or to get properties like the alignments needed for O_DIRECT.  We might be able to conditionally daos_init at that point, or we might need to pull the daos_init over to that part.

We can revisit that idea later.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -271,10 +425,97 @@ int main(int argc, char** argv)
+                    uuid_t cuuid;
+                    rc = dfs_cont_create(src_poh, cuuid, NULL, NULL, NULL);
+                    if (rc != 0) {
+                        MFU_LOG(MFU_LOG_ERR, "Failed to create DFS2 container");
```
Should the errors through this section be fatal?

dsikich replied:
@adammoody I think so, since without a container it will not work at all.

adammoody replied:
Ah, I see.  That makes sense.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -271,10 +425,97 @@ int main(int argc, char** argv)
+        MFU_LOG(MFU_LOG_ERR, "Failed to initialize daos");
+    }
+
+    struct duns_attr_t dattr  = {0}; 
+    struct duns_attr_t ddattr = {0}; 
```
Let's move these just above daos_set_paths if they are only needed there.  Maybe even move them to be local within that function?

adammoody commented on src/dcp/dcp.c:
```diff
@@ -271,10 +425,97 @@ int main(int argc, char** argv)
+    /* connect to DAOS source pool if uuid is valid */
+    if (mfu_src_file->type == DAOS) {
+       /* if DAOS source pool uuid is valid, then set source file type to DAOS */
+        daos_connect(&rank, &src_poh, &src_coh, &src_pool_uuid, &src_cont_uuid, svc); 
```
Do we need error checks on daos_connect?

dsikich replied:
@adammoody the error checking is within the daos_connect function, since it will check for errors when it connects to the pool and opens the container inside the function.

adammoody replied:
What should dcp do if daos_connect fails here?  Do we account for that failure in some other way later on?

dsikich replied:
@adammoody well I've had this happen a few times when I would forgot to start the DAOS servers before running my tests, and DAOS prints and error that says "failed to connect to pool" essentially.

dsikich replied:
It fails within daos when it tries to connect to the pool or open the container, I see what you are saying though, I suppose I can change the error checks in the function to be fatal errors, since I think I have it set to MFU_INFO right now.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -271,10 +425,97 @@ int main(int argc, char** argv)
+                /* create container in same pool */
+                daos_cont_info_t co_info;
+                rc = daos_cont_open(src_poh, dst_cont_uuid, DAOS_COO_RW, &dst_coh, &co_info, NULL);
+	        /* If NOEXIST we create it */
+	        if (rc != 0) {
```
```suggestion
	        if (rc != 0) {
                    /* If NOEXIST we create it */
```

adammoody commented on src/dcp/dcp.c:
```diff
@@ -271,10 +425,97 @@ int main(int argc, char** argv)
+
+    if (mfu_src_file->type == DAOS) {
+        /* DFS is mounted for the source container */
+        rc = dfs_mount(src_poh, src_coh, O_RDWR, &dfs1);
```
Should we have error checks on dfs_mount?

adammoody commented on src/dcp/dcp.c:
```diff
@@ -371,9 +616,62 @@ int main(int argc, char** argv)
+    mfu_file_delete(&mfu_dst_file);
+
+#ifdef DAOS_SUPPORT
+    /* finalize daos */
+    rc = daos_fini();
```
Assuming we add the check for DAOS around the section that contains the daos_init, we'll need the same check down here.

adammoody replied:
Nevermind about this, given our earlier discussion about daos_init.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -120,6 +388,21 @@ int mfu_lstat(const char* path, struct stat* buf)
+        int rc = daos_stat(path, buf, mfu_file);
+        return rc;
+    } else {
+        MFU_ABORT(-1, "File type not known: %s errno=%d (%s)",
+                  path, errno, strerror(errno));
```
errno won't be well defined for this kind of error, but let's report the type value instead
```suggestion
        MFU_ABORT(-1, "File type not known: %s type=%d",
                  path, mfu_file->type);
```
Let's update any other spots like this for unknown mfu_file type values

adammoody commented on src/common/mfu_io.c:
```diff
@@ -561,6 +1095,35 @@ int mfu_rmdir(const char* dir)
+#ifdef DAOS_SUPPORT
+    struct dfs_mfu_t* dirp = calloc(1, sizeof(*dirp));
+    if (dirp == NULL) {
+        return NULL;
```
Let's add an ```errno = ENOMEM;``` here

adammoody commented on src/common/mfu_io.c:
```diff
@@ -603,6 +1197,51 @@ int mfu_closedir(DIR* dirp)
+	                 dirp->ents);
+        if (rc) {
+            MFU_LOG(MFU_LOG_ERR, "dfs_readdir failed (%d %s)", rc, strerror(rc));
+            errno = ENOENT;
```
In the case of an error, do you know whether dfs_readdir will set ```dirp->num_ents=0``` here.  If not, let's add a line here to do that.  That will fail more gracefully in the case where someone does this:
```
dirent* dent = daos_readdir(dirp, mfu_file);

// dfs_readdir returns an error, so daos_readdir sets errno and returns NULL

// user calls daos_readdir again
dent = daos_readdir(dirp, mfu_file);
```
If dirp->num_ents is not set to 0 on that first error, I think the second call will not behave correctly, since it will try to execute these lines which is probably not valid:
```
ret:
    dirp->num_ents--;
    return &dirp->ents[dirp->num_ents];
```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -253,6 +556,34 @@ int mfu_hardlink(const char* oldpath, const char* newpath)
+    int rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+    if (parent != NULL) {
+        rc = dfs_open(mfu_file->dfs, parent, name,
+                      S_IFREG | S_IWUSR | S_IRUSR, O_RDWR | O_CREAT,
```
Can we use the mode settings passed from the user here, or is that not valid?

dsikich replied:
@adammoody I updated this, I can use the mode passed in but also have to pass in a DAOS specific flag with it, S_IFREG | mode.

adammoody replied:
I think adding S_IFREG should be fine.

We should also check whether we can use the ```flags``` parameter that is passed in from the user in place of the hardcoded ```O_RDWR | O_CREAT```.  However, we can do that in a future PR.

dsikich replied:
@adammoody yeah, I already added the flags parameter, that one can be used.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1682,6 +1738,14 @@ static int mfu_copy_file_normal(
 
         /* add bytes to our total (use bytes read,
          * which may be less than number written) */
+#ifdef DAOS_SUPPORT
+        if (mfu_src_file->type == DAOS) {
+            mfu_src_file->offset += num_of_bytes_read;
+        }
+        if (mfu_dst_file->type == DAOS) {
+            mfu_dst_file->offset += num_of_bytes_read;
+        }
+#endif
```
Let's move these offset updates to the daos_read and daos_write functions.  Users calling mfu_file_read and mfu_file_write will expect the file pointers to be updated as part of the call.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -326,6 +697,26 @@ int mfu_close(const char* file, int fd)
+    }
+}
+
+off_t daos_lseek(const char* file, mfu_file_t* mfu_file, off_t pos, int whence)
+{
+    /* noop becuase daos have an lseek */
+    return 0;
+}
+
```
Let's have this set the mfu_file->offset value based on pos/whence values.  For this PR, you could just worry about SEEK_SET and abort if whence is not SEEK_SET.

dsikich replied:
@adammoody ok, this is what I updated to:

```C
    if (whence == SEEK_SET) {
        mfu_file->offset = (daos_off_t)pos;
    } else {
        MFU_ABORT(-1, "whence type not known: %d errno=%d (%s)",
                  whence, errno, strerror(errno));
    }
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1583,76 @@ static int mfu_is_eof(const char* file, int fd)
+        sgl.sg_iovs = &iov;
+        sgl.sg_nr_out = 1;
+        if (mfu_src_file->type == DAOS) {
+            mfu_src_file->offset = (daos_off_t) offset;
```
If you update the daos_lseek to set the mfu_file->offset, then you can remove this line.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1554,50 +1583,76 @@ static int mfu_is_eof(const char* file, int fd)
+        sgl.sg_nr = 1;
+        D_ALLOC(buf, buf_size);
+        D_ASSERT(buf != NULL);
+        d_iov_set(&iov, buf, buf_size);
+        sgl.sg_iovs = &iov;
+        sgl.sg_nr_out = 1;
```
I think we'll eventually want to move setting the iov and/or sql values to the daos_read and daos_write wrappers, like you're setting the size now.  Not sure how much work that might be.  We could push that to a follow up PR if that leads to a ton of work.

dsikich replied:
@adammoody yeah that would take some reworking of the functions, I would prefer to handle that in a separate PR, as this one is getting a bit unwieldy. You can also create Github issues/tickets for DAOS related code fixes or features, and just assign it to me.

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1713,28 +1777,32 @@ static int mfu_copy_file_normal(
 
     /* we don't bother closing the file because our cache does it for us */
-
+#ifdef DAOS_SUPPORT
```
```suggestion

#ifdef DAOS_SUPPORT
```

adammoody commented on src/common/mfu_flist_copy.c:
```diff
@@ -1713,28 +1777,32 @@ static int mfu_copy_file_normal(
+    if (mfu_src_file->type == DAOS || mfu_dst_file->type == DAOS) {
+        D_FREE(buf);
+    }
+#endif
```
```suggestion
#endif

```

adammoody commented on src/common/mfu_io.c:
```diff
@@ -326,6 +697,26 @@ int mfu_close(const char* file, int fd)
+{
+    if (mfu_file->type == POSIX) {
+        int rc = mfu_close(file, mfu_file->fd);
+        return rc;
```
if rc == 0 here, let's set mfu_file->fd back to its initial value (-1?), with a similar change for daos.  That might help use catch other bugs.  But only on success.  On error, let's leave it as it, because they may still need to do something with it in that case.

adammoody commented on src/dcp/dcp.c:
```diff
@@ -371,9 +619,62 @@ int main(int argc, char** argv)
+#ifdef DAOS_SUPPORT
+    /* finalize daos */
+    rc = daos_fini();
+#endif
```
```suggestion
#endif

```

adammoody commented:
There are some cases where dcp can call mfu_unlink.  We'll probably want an mfu_file_unlink for that, but we can do that on a future PR.

dsikich commented:
@adammoody ok, I have pushed my changes based on your comments and pushed all of the suggested changes you added. Is it possible to get this merged in by end of day tomorrow (Friday)? I am not seeing any glaring issues/bugs and this PR is getting unwieldy. I think for small updates it would be easier to submit smaller PR's where its easier to track. You can feel free to add feature requests as Github issues and just tag me on it for DAOS ones.

dsikich commented:
@adammoody saw that there was a build error. I missed an ifdef for DAOS_SUPPORT where I was setting the mfu_file object back to NULL. It is good to know travis is checking the build when DAOS is not present! Build should be good now.

adammoody commented:
Sure.  Let's get this one merged and we can finish things up in smaller PRs.

Thanks for all of your work on this one!

dsikich commented:
@adammoody thank you for all of your feedback! I have a draft of some documentation I will try to submit next week as well.

adammoody commented:
Sounds good.  I created a few issues for some of the remaining items while they were fresh in my head.  I know there was a lot of work in adding this mfu_file abstraction, but it will make it easier to convert over the remaining tools to have DAOS equivalents.  It will pay off well.
